File: exercism/haskell/allergies/src/Allergies.hs
module Allergies (Allergen(..), allergies, isAllergicTo) where

import Data.Bits (testBit)

data Allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats
              deriving (Bounded, Enum, Eq)

allergies :: Int -> [Allergen]
allergies score = filter (`isAllergicTo`score) [minBound .. maxBound]

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo = flip testBit . fromEnum

File: exercism/ocaml/allergies/allergies.ml
type allergen =
  | Eggs
  | Peanuts
  | Shellfish
  | Strawberries
  | Tomatoes
  | Chocolate
  | Pollen
  | Cats

let a2i = function
  | Eggs -> 1
  | Peanuts -> 2
  | Shellfish -> 4
  | Strawberries -> 8
  | Tomatoes -> 16
  | Chocolate -> 32
  | Pollen -> 64
  | Cats -> 128

let allergic_to n a = 0 != Int.logand n @@ a2i a

let allergies n =
  List.filter (allergic_to n)
    [
      Eggs; Peanuts; Shellfish; Strawberries; Tomatoes; Chocolate; Pollen; Cats;
    ]
